insecure/adversary/attacker_test.go

```diff
@@ -0,0 +1,195 @@
+// It also generates and returns the corresponding application-layer event of that message, which is sent between the attacker and the
+// orchestrator.
+func messageFixture(t *testing.T, codec network.Codec) (*insecure.Message, *insecure.Event) {
+	// fixture for content of message
+	originId := unittest.IdentifierFixture()
+	targetIds := unittest.IdentifierListFixture(10)
+	targets := uint32(3)
+	protocol := insecure.Protocol_MULTICAST
+	channel := network.Channel("test-channel")
+	content := &message.TestMessage{
+		Text: fmt.Sprintf("this is a test message: %d", rand.Int()),
+	}
+
+	// encodes event to create payload
+	payload, err := codec.Encode(content)
+	require.NoError(t, err)
+
+	// creates message that goes over gRPC.
+	m := &insecure.Message{
+		ChannelID: "test-channel",
+		OriginID:  originId[:],
+		Targets:   targets,
+		TargetIDs: flow.IdsToBytes(targetIds),
+		Payload:   payload,
+		Protocol:  protocol,
+	}
+
+	// creates corresponding event of that message that
+	// is sent by attacker to orchestrator.
+	e := &insecure.Event{
+		CorruptedId: originId,
+		Channel:     channel,
+		Content:     content,
+		Protocol:    protocol,
+		TargetNum:   targets,
+		TargetIds:   targetIds,
+	}
+
+	return m, e
+}
+
+// messageFixtures creates and returns randomly generated gRCP messages and their corresponding protocol-level events.
+// The messages are sent between a corruptible conduit and the attacker.
+// The events are the corresponding protocol-level representation of messages.
+func messageFixtures(t *testing.T, codec network.Codec, count int) ([]*insecure.Message, []*insecure.Event) {
+	msgs := make([]*insecure.Message, count)
+	events := make([]*insecure.Event, count)
+
+	for i := 0; i < count; i++ {
+		m, e := messageFixture(t, codec)
+
+		msgs[i] = m
+		events[i] = e
+	}
+
+	return msgs, events
+}
+
+// withAttacker is a test helper that creates and starts an attacker server, prepares a set of message fixtures and runs the given scenario
+// against the attacker and generated messages.
+func withAttacker(
+	t *testing.T,
+	messageCount int, // number of messages received by attacker from corruptible conduits while playing scenario.
+	scenario func(*testing.T, *mockinsecure.AttackOrchestrator, insecure.Attacker_ObserveClient, []*insecure.Message, []*insecure.Event)) {
+
+	codec := cbor.NewCodec()
+	orchestrator := &mockinsecure.AttackOrchestrator{}
+	// mocks start up of orchestrator
+	orchestrator.On("Start", mock.AnythingOfType("*irrecoverable.signalerCtx")).Return().Once()
+
+	attacker, err := NewAttacker(unittest.Logger(), attackerAddress, codec, orchestrator)
+
+	require.NoError(t, err)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	attackCtx, errChan := irrecoverable.WithSignaler(ctx)
+
+	go func() {
+		select {
+		case err := <-errChan:
+			t.Error("attacker startup encountered fatal error", err)
+		case <-ctx.Done():
+			return
+		}
+	}()
+
+	// starts attacker
```

Comment:
can you add comments here to better explain what exactly is happening?

Reply:
The logic is quite self-explanatory and common across our codebase. So, let's discuss it in more context here instead of adding redundant comments to the code that may stale over the time:

Starting the attacker starts its sub-components. In the Flow codebase, we define the notion of [`Component`](https://github.com/onflow/flow-go/blob/57f89d4e96259f08fe84163c91ecd32484401b45/module/component/component.go#L202) as an abstraction of a _control_ logic. At the very broad level, the protocol logic is decomposed into self-contained `Component`s each with a certain responsibility. Smaller `Component`s are composed together and shaped into bigger ones with more general responsibilities. Examples of such in our BFT testing implementation are `Attacker` and `Orchestrator`. The `Orchestrator` implements a certain type of attack (e.g., the wintermute), and is _composed_ into the bigger `Attacker` component that is responsible for managing the orchestrator together with a bunch of corrupted conduit factories. 

In such a design, we cannot let components commence their work at the initialization time. The reason is some components depend on others, and some are composed into bigger ones. There is no guarantee provided by Golang that all components are initialized together even if we attempt to run them concurrently. This provides motivations to decouple the components _initialization_ from _startup_. 

One may see the _startup_ as triggering a tree of events. The biggest component (i.e., the node) starts first. As part of its startup, it starts its constituent components, which in turn each component starts its subcomponents, etc. 

But how do we make sure that a component is started successfully? by looking at its `Ready` method. The `Ready` returns a channel that gets closed when the component is (and hence all its sub-component are) ready. That is why we do `attacker.Start(...)` and then wait for `attacker.Ready`'s returned channel to get closed. 

But how we can make sure that a fault in the attacker's startup does not block the entire test? the answer is the `RequireClosesBefore` helper function of our unit-testing. It fails the entire test if the given channel is not closed prior to a deadline. So the following means start the attacker and if it fails to get started completely within 1 second, fail the entire test. 

```
	attacker.Start(attackCtx)
	unittest.RequireCloseBefore(t, attacker.Ready(), 1*time.Second, "could not start attacker on time")
```